src/v/net/conn_quota.cc

```diff
@@ -0,0 +1,440 @@
+
+    allowance.max = new_limit;
+    if (in_use >= allowance.max) {
+        vlog(
+          rpc::rpclog.trace,
+          "Connection count limit {} decreased below current ({}) for {}",
+          allowance.max,
+          in_use,
+          addr);
+        allowance.available = 0;
+    } else {
+        allowance.available = allowance.max - in_use;
+    }
+
+    // If the allowance might have had borrowed units on other
+    // nodes, then we must reclaim them in case of a decrease
+    // to the limit.
+    if (was_dirty_decrease) {
+        ssx::spawn_with_gate(_gate, [this, addr]() {
+            auto& allowance = get_home_allowance(addr);
+            return reclaim_to(allowance, addr, true);
+        });
+    }
+}
+
+conn_quota::home_allowance&
+conn_quota::get_home_allowance(ss::net::inet_address addr) {
+    assert_on_home(addr);
+
+    if (addr == ss::net::inet_address{}) {
+        return total_home;
+    } else {
+        auto found = ip_home.find(addr);
+        if (found != ip_home.end()) {
+            return found->second;
+        } else {
+            auto [iter, created] = ip_home.insert(std::make_pair(
+              addr,
+              home_allowance{
+                .max = _cfg.max_connections_per_ip().value(),
+                .available = _cfg.max_connections_per_ip().value()}));
+            return iter->second;
+        }
+    }
+}
+
+conn_quota::remote_allowance&
+conn_quota::get_remote_allowance(ss::net::inet_address addr) {
+    if (addr == ss::net::inet_address{}) {
+        return total_remote;
+    } else {
+        auto found = ip_remote.find(addr);
+        if (found != ip_remote.end()) {
+            return found->second;
+        } else {
+            auto [iter, created] = ip_remote.insert(
+              std::make_pair(addr, remote_allowance{}));
+            return iter->second;
+        }
+    }
+}
+
+/**
+ * Called on the home shard for `addr`.  Dispatch reclaim requests
+ * to all other shards, hopefully they had some borrowed tokens
+ * that we can claw back.
+ */
+ss::future<> conn_quota::reclaim_to(
+  conn_quota::home_allowance& allowance,
+  ss::net::inet_address addr,
+  bool one_time) {
+    // Note: require that allowance is pointer-stable for this lock
+    // to be safely freed later (i.e. the ip_home container must
+    // provide pointer stability)
+
+    auto locked = allowance.reclaim_lock.get_units();
+    if (allowance.reclaim) {
+        // We are already in reclaim mode: remote allowances will
+        // not be holding any units belong to us, so don't waste
+        // time looking.
+        co_return;
+    }
+
+    allowance.reclaim = true;
+    uint32_t total_released = 0;
+    for (ss::shard_id i = 0; i < ss::smp::count; ++i) {
+        if (i == ss::this_shard_id()) {
+            continue;
+        }
+
+        total_released += co_await container().invoke_on(
+          i, [addr, one_time](conn_quota& cq) {
+              return cq.reclaim_from(addr, one_time);
```

Comment:
I don't think this is particularly susceptible, but one thing to watch out for are shutdown races with code like this. As an example:

Say this runs on core 5:

```
    if (was_dirty_decrease) {
         ssx::spawn_with_gate(_gate, [this, addr]() {
             auto& allowance = get_home_allowance(addr);
             return reclaim_to(allowance, addr, true);
         });
     }
```

which results in the loop this comment is attached to running `reclaim_from` on all other cores. if shutdown of the service were to call stop on core 5 first, then the gate would wait to drain work from the other cores. but if stop were called on any other core no gate would protect the reclaim_from function from accessing an instance on conn_quota that has been destroyed.

it's probably unlikely here since reclaim_from doesn't block, but the service also strikes me as something that could potentially have a lot of outstanding cross-core work during shutdown.

Reply:
I was just poking around in seastar and noticed that stopping the service is phased so that there is a barrier between global stop and any service being freed. Maybe this isn't an issue anymore; I just remembered fixing something similar in the controller a long time ago, though.

Reply:
It certainly reads to me that all the stop() calls happen before any of the deletes happen in `sharded<Service>::stop()`

The more involved case  seems to be the converse, that there is no guarantee that all the deletes happen before the outer sharded::stop call completes (because each shard's instance is a shared_ptr that might have other refs), and they have `async_sharded_service` for people who want that extra guarantee of destruction during stop.  Not that that is relevant to the code here, I was just reading up.